### PR TITLE
Analytics endpoint

### PR DESCRIPTION
**Pull request by zoek1**
**TODO**
-  Implement aggregate functions for fields
   - [x] fulfillments_pending_acceptance
   - [x] fulfillment_acceptance_rate
   - [x] bounty_fulfilled_rate
   - [x] avg_fulfiller_acceptance_rate
   - [x] avg_fulfillment_amount
   - [x] total_fulfillment_amount

- [x] Add schema field (gitcoin or bounties network)

Seed the bounties time line:
```shell
$ docker-compose exec bounties_api python manage.py timeline_generator
```
Demo request - `http://localhost:8000/analytics/?since=1992-04-23`
timeline bounties django rest framework 2

## Filtering by fields 
`http://localhost:8000/analytics/?since=2018-01-01&fields=bounties_issued,date`
timeline bounties django rest framework filter by field

## Filtering by fields 
`localhost:8000/analytics/?since=2018-01-01&bounties_issued__lt=3`
timeline bounties django rest framework filter by value

### CODE REVIEW

File: bounties_api/analytics/management/commands/timeline_generator.py [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+
+    return stages
+
+
+def generate_timeline(time_frame):
+    date = time_frame[1]
+
+    bounties_state_frame = BountyState.objects.filter(change_date__lte=time_frame[1])
+    bounty_created_frame = Bounty.objects.filter(bounty_created__range=time_frame)
+    fulfillment_accepted_frame = Fulfillment.objects.filter(accepted_date__lte=time_frame[1])
```
Comment (villanuevawill) [inline bounties_api/analytics/management/commands/timeline_generator.py]: does this grab the null values or no? This is safe?

File: bounties_api/analytics/management/commands/timeline_generator.py [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+
+
+    bounties_issued = get_bounties_issued(bounties_state_frame)
+
+    fulfillments_submitted = get_fulfillments_submitted(fulfillment_submitted_frame)
+    fulfillments_accepted = get_fulfillments_accepted(fulfillment_accepted_frame)
+
+    fulfillment_acceptance_rate = get_fulfillment_acceptance_rate(fulfillment_submitted_frame, date)
+
+    bounty_fulfilled_rate = get_bounty_fulfilled_rate(fulfillment_submitted_frame,
```
[the remark this answers is not in the document]
Comment (villanuevawill) [reply bounties_api/analytics/management/commands/timeline_generator.py]: Also, we probably want to not include bounties from the count that never were in an active state. ie. a bounty that was in draft forever, or was killed after being in draft.

File: bounties_api/analytics/management/commands/timeline_generator.py [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+def get_fulfillments_accepted(time_frame):
+    return time_frame.count()
+
+
+def get_fulfillment_acceptance_rate(time_frame, accepted_date=datetime.now()):
+    return time_frame.filter(accepted_date__lte=accepted_date).count() / time_frame.count()
+
+
+def get_bounty_fulfilled_rate(time_frame, bounties):
+    return time_frame.distinct('bounty').count() / bounties.count()
```
[the remark this answers is not in the document]
Comment (villanuevawill) [reply bounties_api/analytics/management/commands/timeline_generator.py]: Ah makes sense - the bounty state time filter grabs the cumulative.

File: bounties_api/analytics/management/commands/timeline_generator.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+    bounty_created_frame = Bounty.objects.filter(bounty_created__range=time_frame)
+    fulfillment_accepted_frame = Fulfillment.objects.filter(accepted_date__lte=time_frame[1])
+    fulfillment_submitted_frame = Fulfillment.objects.filter(fulfillment_created__lte=time_frame[1])
+
+    stages = build_stages(bounties_state_frame)
+
+
+    bounties_issued = get_bounties_issued(bounties_state_frame)
+
+    fulfillments_submitted = get_fulfillments_submitted(fulfillment_submitted_frame)
```
Comment (villanuevawill) [inline bounties_api/analytics/management/commands/timeline_generator.py]: any chance we can also add an additional field for these 3 metrics:

1. bounty_issued
2. fulfillments_submitted
3. fulfillments_accepted

Currently ^^ those are cumulative numbers.
Could we also include non-cumulative. Maybe we should call the ones above fulfillments_accepted_cum and fulfillments_accepted as an example.

File: bounties_api/analytics/management/commands/fake_timeline.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
+from django.core.management import BaseCommand
```
Comment (villanuevawill) [inline bounties_api/analytics/management/commands/fake_timeline.py]: So far lookinmg really good!

**Comment by villanuevawill** [comment]
And then of course the schema it also needs to be filtered by... one of those being 'ALL'

File: bounties_api/analytics/management/commands/timeline_generator.py (hunk 4 above)
Comment (zoek1) [reply bounties_api/analytics/management/commands/timeline_generator.py]: Sure, such fields was added at ed89f08e8 :smile:

File: bounties_api/analytics/management/commands/timeline_generator.py (hunk 1 above)
Comment (zoek1) [reply bounties_api/analytics/management/commands/timeline_generator.py]: `bounty_created_frame` was deleted in favor to access bounties through BountyState. And nop, such queries doesn't grab null values

**Comment by zoek1** [comment]
the option to filtering is done @villanuevawill at 9fb9646.

**Comment by zoek1** [comment]
@villanuevawill now supports filtering by schema. Current schemas are getted from the bounties schema fields. At this moments are `['None', 'gitcoinBounty', 'standardSchema']`. Also i added `'all'` to schemas.

**Example query by schema**:
`http://localhost:8000/analytics/?since=1992-04-23&schema=gitcoinBounty`

**Comment by villanuevawill** [comment]
@zoek1 can you make it so none defaults to the bounty network schema?

**Comment by zoek1** [comment]
@villanuevawill  `standardSchema` is now the default schema for bounties with the value `None` of its schema.

File: bounties_api/analytics/management/commands/timeline_generator.py [hunk 6]
```diff
@@ -0,0 +1,265 @@
...
+
+    return bounty_frame
+
+
+class Command(BaseCommand):
+    def handle(self, *args, **options):
+        needs_genesis = not BountiesTimeline.objects.all().count()
+        schemas_query = BountyState.objects.distinct('bounty__schemaName')
+        schemas = [schema.bounty.schemaName for schema in schemas_query if schema.bounty.schemaName] + ['all']
+        print(schemas)
```
Comment (villanuevawill) [inline bounties_api/analytics/management/commands/timeline_generator.py]: @zoek1 use the logger here please. logger.info

Comment (zoek1) [reply bounties_api/analytics/management/commands/timeline_generator.py]: Sorry, such print will ve deleted 🙈. i forgot delete it

File: bounties_api/analytics/management/commands/timeline_generator.py [hunk 7]
```diff
@@ -0,0 +1,265 @@
...
+    for bounty_id in unique_bounties:
+        bounty_states = time_frame.filter(bounty=bounty_id).order_by('-change_date')
+        bounty_stages[bounty_id] = map(lambda b: b.bountyStage, bounty_states)
+        current_stage = bounty_states.first().bountyStage
+        stages[current_stage] = stages[current_stage] + 1
+
+    return stages, bounty_stages
+
+
+def get_noise_bounties(bounties):
```
Comment (villanuevawill) [inline bounties_api/analytics/management/commands/timeline_generator.py]: very elegant.

File: bounties_api/analytics/views.py [hunk 8]
```diff
@@ -1,3 +1,48 @@
...
+
+
+                serialized = BountiesTimelineSerializer(bounties_timeline.qs, many=True, context={'request': request})
+
+                return Response(serialized.data)
+        except ValueError:
+            pass
+
+        res = {"error": 400, "message": "The fields since & until needs being formated as YYYY-MM-DD"}
+        return Response(json.dumps(res), status=status.HTTP_200_OK)
```
Comment (villanuevawill) [inline bounties_api/analytics/views.py]: excellent work - I'm deploying this now to try out.

File: bounties_api/analytics/management/commands/timeline_generator.py [hunk 9]
```diff
@@ -0,0 +1,264 @@
...
+                                    bounty_expired=stages[EXPIRED_STAGE],
+                                    bounty_dead=stages[DEAD_STAGE],
+                                    schema=schema)
+
+    return bounty_frame
+
+
+class Command(BaseCommand):
+    def handle(self, *args, **options):
+        needs_genesis = not BountiesTimeline.objects.all().count()
```
Comment (villanuevawill) [inline bounties_api/analytics/management/commands/timeline_generator.py]: needs_genesis logic and also last_update logic can introduce a weird condition if there is a record for one schema but not the others (say it fails).

This means if standardSchema is succssful, but then the gitcoinSchema fails for that day... we're out of luck.

**Comment by villanuevawill** [comment]
@zoek1 order_by date needs to be set as a default here.

File: bounties_api/analytics/management/commands/timeline_generator.py (hunk 9 above)
Comment (zoek1) [reply bounties_api/analytics/management/commands/timeline_generator.py]: Now genesis is based on the current state of each schema in `BountiesTimeline` 6d8055bc0030b21d.

**Comment by amites** [comment]
@zoek1 probably wanna drop a note in origin issue -- gitbot has already booted ya from the bounty

**Comment by villanuevawill** [comment]
@zoek1 just consolidate the migrations please to one file... then this will look ready to merge.

**Comment by zoek1** [comment]
@villanuevawill squashed migrations done :smile:

**Pull request merged**